CW-1250: Refactor L2 Chains Code

Blazebrain opened the pull request:
Issue Number (if Applicable): Fixes #

# Description

Please include a summary of the changes and which issue is fixed / feature is added. 

# Pull Request - Checklist  

- [ ] Initial Manual Tests Passed
- [ ] Double check modified code and verify it with the feature/task requirements
- [ ] Format code
- [ ] Look for code duplication
- [ ] Clear naming for variables and methods
- [ ] Manual tests in accessibility mode (TalkBack on Android) passed

Blazebrain committed:
feat(evm): add ChainConfig model and related classes

- Add ChainConfig class for immutable EVM chain configuration
- Add ChainCapabilities class for chain feature flags
- Add FeeType enum and FeeModel class for fee configuration

Blazebrain committed:
feat(evm): add EvmChainRegistry for centralized chain management

- Add singleton EvmChainRegistry class with chain configuration storage
- Initialize with Ethereum, Polygon, Base, and Arbitrum chains
- Add mappings for WalletType, tag, and CAIP-2 to chainId lookups
- Provide lookup methods for chain configurations by various identifiers
- Support for querying available chains and registered chain IDs

Blazebrain committed:
feat(evm): add EVMChainClientFactory and consolidate chain clients

- Create EVMChainClientFactory for creating chain-specific clients based on chainId
- Move all chain clients (EthereumClient, PolygonClient, BaseClient, ArbitrumClient) into cw_evm/lib/clients/ folder
- Refactor EVMChainClient from abstract to concrete class with default implementations
- Add default implementations for fetchTransactions, fetchInternalTransactions, and prepareSignedTransactionForSending
- Update all client classes to use super(chainId: X) constructor pattern
- Factory returns EVMChainClient directly for unregistered chains instead of throwing
- Remove dependencies on cw_ethereum, cw_polygon, cw_base, cw_arbitrum packages from pubspec.yaml
- Consolidate all EVM chain client implementations in one location for easier maintenance

malik1004x committed:
add ui for evm switcher on eth

malik1004x committed:
add inputs for evm network list edit switches

Blazebrain committed:
refactor(evm): Make EVMChainWallet concrete and unify transaction classes
- Make EVMChainWallet and EVMChainWalletBase concrete classes (removed abstract)
- Implement all previously abstract methods in EVMChainWallet
- Make EVMChainTransactionInfo and EVMChainTransactionHistory concrete
- Reorganize default tokens into tokens/ folder:
- Refactor EVMChainDefaultTokens to import and use token classes from tokens/ folder
- Move clients to clients/ folder:
- Move evm_chain_formatter.dart to utils
- Update chain-specific transaction history classes to pass walletType to fromJson
- Update old wallet classes (EthereumWallet, PolygonWallet, etc.) to pass walletType
- Maintain compatibility with existing per-chain wallet classes

This refactoring enables the unified EVMChainWallet architecture while maintaining
full backward compatibility with existing per-chain wallet implementations.

Blazebrain committed:
feat(evm): Add chain selection capability to EVMChainWallet
- Add selectedChainId observable field to track currently selected chain
- Add selectChain(chainId) action method to switch between EVM chains
- Add selectedChainConfig computed getter for accessing chain configuration
- Initialize selectedChainId from client.chainId in constructor
- Add _getClientForCurrentChain() helper method for future use

This enables chain switching functionality while maintaining full backward compatibility. The selectedChainId is initialized from the existing client, and all existing methods continue to work unchanged. Future increments will update methods to use selectedChainId internally.

Blazebrain committed:
refactor(evm): Update client immediately on chain selection

- Update selectChain() to immediately create new client for selected chain
- Remove _getClientForCurrentChain() helper method
- Simplify all methods to use _client directly instead of helper calls
- Client is now always in sync with selectedChainId

This improves code simplicity and performance by eliminating repeated
client checks. The client is updated synchronously when selectChain()
is called, ensuring state consistency.

Blazebrain committed:
featr(evm): Add unified evm proxy setup and fix chain initialization

- Add unified Evm proxy
- Add evm.dart to .gitignore for generated proxy
- Add generateEVM() to configure.dart for proxy generation
- Fix selectedChainId to initialize from registry based on walletInfo.type

Ensures selectedChainId always matches wallet type on initialization
and adds unified proxy infrastructure.

Blazebrain committed:
refactor(evm): Update backward compat helpers to use registry via proxy

- Add registry helper methods to unified EVM proxy
- Update all lib/ files to use proxy instead of direct cw_evm imports
- Enforce proxy pattern: no direct imports from cw_evm in main app
- Update configure.dart to include registry methods in generated proxy

Maintains backward compatibility while centralizing chain data access
through the unified proxy pattern.

Blazebrain committed:
feat(evm): Add chain selection UI to dashboard

- Add chain selection methods to DashboardViewModel (isEVMWallet,
  availableChains, currentChain, selectChain)
- Add chain dropdown widget to balance page for EVM wallets
- Dropdown shows current chain and allows switching between all
  registered EVM chains
- Only visible for EVM-compatible wallets
- Add chain selection methods to EVM proxy (getAllChains,
  getCurrentChain, selectChain)

Users can now switch between EVM chains (Ethereum, Polygon, Base,
Arbitrum) directly from the dashboard without creating separate wallets. All chain-related code follows the established proxy pattern.

Blazebrain committed:
feat(evm): Add persistence for selected chain ID

- Add initialChainId parameter to EVMChainWallet constructor
- Save selectedChainId to wallet JSON in toJSON() method
- Load selectedChainId from JSON in open() method
- Handle backward compatibility: wallets without saved chain ID use
  wallet type's default chain ID
- Client is created with saved chain ID when opening wallet

Selected chain preference is now persisted across app sessions. Users will see their previously selected chain when reopening wallets.

Blazebrain committed:
chore: Minor cleanup

Blazebrain committed:
refactor(evm): Remove old per-chain packages, consolidate to unified cw_evm

- Delete cw_ethereum, cw_polygon, cw_base, cw_arbitrum packages
- Remove old proxy files (lib/ethereum/, lib/polygon/, lib/base/, lib/arbitrum/)
- Update all view models to use unified evm! proxy
- Move DEuro functionality to cw_evm/lib/deuro/
- Remove dependencies from pubspec.yaml and configure.dart
- Remove walletType parameter from EVM proxy methods

All EVM chains now use unified cw_evm package through single proxy interface. lib/ only depends on cw_evm, simplifying architecture and enabling easy addition of new L2 chains.

BREAKING CHANGE: Old per-chain packages removed

Blazebrain committed:
refactor(evm): Update wallet handling to use unified EVM proxy

- Replace individual chain imports (Ethereum, Polygon, Base, Arbitrum) with a single EVM proxy import.
- Refactor wallet-related methods across various view models to utilize the new evm proxy for various usecases.

This change enhances code maintainability and prepares the codebase for the addition of new EVM chains.

Blazebrain committed:
refactor(evm): Update model generator and enhance EVM classes

- Modify model_generator.sh to remove previous chain folders from the loop.
- Change EVMChainTransactionHistoryBase and EVMChainWalletBase classes to be abstract.
- Update import paths in configure.dart to reflect the new structure of the cw_evm package.

malik1004x committed:
add evm switcher icons to pubspec_base

malik1004x committed:
change evm switcher to sliding animation

malik1004x committed:
formatting fix

malik1004x committed:
dpi agnostic width

Blazebrain committed:
feat(evm): Enhance chain handling and transaction history management

- Introduce getCurrentChainId function to EVMChainTransactionHistoryBase for dynamic transaction history file naming based on the current chain ID.
- Update EVMChainTransactionInfo to include chainId, ensuring accurate transaction data representation.
- Refactor EVMChainWalletBase to support automatic node connection and transaction history loading upon chain selection.
- Implement saveBackup and restoreWalletFilesFromBackup methods in EVMChainWalletService to manage wallet backups based on wallet type.
- Update various methods across the codebase to accommodate the new chain handling logic, improving overall functionality and user experience.

Blazebrain committed:
chore: Cleanups and updates

Blazebrain committed:
feat: add WalletType.evm and deprecate individual EVM chain types

Add unified WalletType.evm enum value for all EVM-compatible chains.
Deprecate old individual EVM types (ethereum, polygon, base, arbitrum)
with guidance to use WalletType.evm for new code.

Update serialization/deserialization functions to support WalletType.evm:
- serializeToInt() returns 18 for WalletType.evm
- deserializeFromInt() handles case 18
- walletTypeToString() and walletTypeToDisplayName() return 'EVM'
- cryptoCurrencyToWalletType() maps all EVM currencies to WalletType.evm

Old EVM types remain functional for backward compatibility with existing
wallets. This is the foundation for unifying EVM chain management.

Blazebrain committed:
feat: update EVM detection functions to support WalletType.evm

Update all EVM detection and chain helper functions to support the unified WalletType.evm while maintaining backward compatibility with old EVM types.

All functions require chainId parameter when used with WalletType.evm, ensuring proper chain-specific behavior. Old EVM types continue to work without changes for backward compatibility.

Blazebrain committed:
feat: add currency mapping support for unified EVM wallet type

- Update currency mapping to support WalletType.evm with chainId-based currency resolution.
- Override EVMChainWallet.currency getter to use selectedChainId from registry.
- Replace walletTypeToCryptoCurrency calls with wallet.currency in view models.
- Add WalletType.evm cases to switch statements and consolidate EVM transaction list item methods.
- Add explorer URL method to EVM interface

Blazebrain committed:
feat: Update wallet operations with unified WalletType.evm

- Update wallet service to always create new wallets and restores with WalletType.evm, defaulting to Ethereum.
- Preserve old wallet types when opening existing wallets. Update wallet instance creation and opening logic to handle unified EVM type with chainId-based config lookup.

Blazebrain committed:
Add WalletType.evm support across all switch statements
- Update all switch statements and conditionals to include WalletType.evm alongside existing EVM types.
- Add chainId-based node selection support for unified EVM wallets. - Fix redundant code in exchange view model.
- Ensure backward compatibility with old EVM wallet types.

Blazebrain committed:
feat(evm): Implement chainId-based node management for WalletType.evm

- Add chainId-based node storage for unified EVM wallets, enabling separate node preferences per chain.
- Replace unsafe dynamic casts with type-safe getSelectedChainId method. Update all node retrieval and switching logic to support chainId-based selection.

Blazebrain committed:
feat: Update UI components to show unified EVM wallet option

- Filter out old EVM wallet types (ethereum, polygon, base, arbitrum) from wallet creation and restore UI, showing only unified WalletType.evm option.
- Update wallet type generation to include WalletType.evm instead of individual
chain types.
- Update configure.dart to generate WalletType.evm in availableWalletTypes
- Add filtering checks to prevent old EVM types from appearing

Blazebrain committed:
fix: Fix WalletType.evm support and complete remaining edge cases

Add missing WalletType.evm cases to switch statements to resolve compilation
errors. Complete remaining unification items:

- Add WalletType.evm support to node, wallet utils, and view models
- Update EVMChainDefaultTokens and EVMChainUtils to support chainId
- Fix token initialization and priority fee calculations for WalletType.evm
- Update payment view model to handle WalletType.evm with chainId

All identified edge cases from the unification plan are now complete.

Blazebrain committed:
fix: Fix network switching issues for EVM wallets
- Fix null check errors when creating and restoring an EVM wallet
- Fix null check errors for balance getter when switching networks
- Fix fiat amount display for native currencies in transaction history

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
Merge branch 'cw-1300-evm-switcher' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
refactor: Simplify EVM wallet chain selection in CryptoBalanceWidget

- Removed the old dropdown for chain selection and replaced it with a more integrated EvmSwitcher dialog.
- Updated EvmSwitcher to accept a list of available chains and handle chain selection more efficiently.
- Enhanced EvmSwitcherDataItem to include chainId for better identification of chains.
- Improved the overall UI flow for EVM wallet interactions.

Blazebrain committed:
refactor: Enhance EVM wallet handling and chainId integration

- Simplified chainId checks across various components to ensure consistent handling for WalletType.evm.
- Updated EVMChainDefaultTokens and EVMChainUtils checks to allow all EVM wallets for chainId operations
- Handled backward compatibility for older wallet types in transaction info

Blazebrain committed:
refactor: More enhancement and backward compatibility fixes for EVM Wallet

- Introduced a new method to retrieve cryptocurrency by chainId, improving the handling of WalletType.evm.
- Updated transaction history management to filter transactions based on the current chainId
- Refactored EVMChainTransactionInfo to accept chainId directly
- Modified utility functions in EVMChainUtils and TokenUtilities for better chainId handling

Blazebrain committed:
refactor: Streamline EVM token retrieval and enhance chainId handling

Blazebrain committed:
feat: Introduce comprehensive guide for adding new EVM L2 networks

- Added a detailed documentation file outlining the steps to integrate new EVM-compatible L2 networks into Cake Wallet.
- Included prerequisites, architecture overview, and a step-by-step guide covering chain configuration, native currency addition, default tokens, and node setup.
- Emphasized the unified architecture for EVM chains and backward compatibility considerations.
- Removed the outdated guide on adding new L2 networks to streamline documentation.

Blazebrain committed:
Merge remote into local

Blazebrain committed:
feat: Reposition EVM switcher button on dashboard page

OmarHatem28 committed:
just adding reminders for the refactoring [skip ci]

OmarHatem28 committed:
more todos [skip ci]

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
feat: enhance wallet functionality with chainId support

- Added chainId to `WalletBase` and updated currency method to use it.
- Modified EvmChainRegistry to initialize registry on call.
- Updated various components to pass chainId where necessary for currency conversions.
- Improved handling of chain IDs in wallet-related view models and UI components for better compatibility with EVM networks.

Blazebrain committed:
Updating

Blazebrain committed:
fix: remove default chainId fallback in payment confirmation widget

OmarHatem28 commented on lib/view_model/contact_list/contact_list_view_model.dart:
```diff
@@ -44,7 +47,11 @@ abstract class ContactListViewModelBase with Store {
+                info.type,
+                chainId: info.type == WalletType.evm ? evm!.getChainIdByWalletType(info.type) : null,
+              ),
+              walletType: info.type,
```
why are you defaulting to ethereum chain id?

OmarHatem28 commented on lib/view_model/dashboard/balance_view_model.dart:
```diff
@@ -66,6 +66,16 @@ abstract class BalanceViewModelBase with Store {
+      return null;
+    }, (_) {
+      wallet.balance;
+    });
```
what's the point of this?

OmarHatem28 commented on lib/view_model/send/fees_view_model.dart:
why is arbitrum missing from here?

Blazebrain replied:
Arbitrum does not support priority fees, it processes based on a first in first out (FIFO) basis

OmarHatem28 commented on lib/view_model/anon_invoice_page_view_model.dart:
```diff
@@ -182,8 +182,8 @@ abstract class AnonInvoicePageViewModelBase with Store {
-    selectedCurrency = walletTypeToCryptoCurrency(_wallet.type);
-    cryptoCurrency = walletTypeToCryptoCurrency(_wallet.type);
+    selectedCurrency = walletTypeToCryptoCurrency(_wallet.type, chainId: _wallet.chainId);
+    cryptoCurrency = walletTypeToCryptoCurrency(_wallet.type, chainId: _wallet.chainId);
```
sometimes you are using `wallet.currency` and someothers you are using `walletTypeToCryptoCurrency`, let's unify one across all places, so if we will make any modifications it will be in one place

Blazebrain committed:
feat: Adapt anypay to new unified flow WIP

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
fix: Issues with handling non evm swaps for anypay

Blazebrain committed:
fixes to flow structure

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
Add persistence to hidden chains on switcher

Blazebrain committed:
refactor: simplify equality operator in Erc20Token class

Blazebrain committed:
feat: add excluded arbutrum tokens for chainflip

Blazebrain committed:
chore: Remove print statement [skip ci]

Blazebrain committed:
refactor: dispaly QR image display for evm based on selected chain ID

Blazebrain committed:
feat: enhance EVM wallet compatibility check for raw address input

Blazebrain committed:
chore: Update doc to reflect changes [skip ci]

Blazebrain committed:
Update

Blazebrain committed:
feat: Integrate Blink for EVM wallets for anti-sandwich attacks.

- Add toggle in privacy settings to enable/disable
- When enabled, route all evm broadcast rpc calls via Blink

Blazebrain committed:
feat: Integrate Blink for EVM wallets for anti-sandwich attacks.

- Add toggle in privacy settings to enable/disable
- When enabled, route all evm broadcast rpc calls via Blink

Blazebrain committed:
Merge branch 'CW-1250-Refactor-L2-Chains-Code' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
feat: Integrate Blink for EVM wallets for anti-sandwich attacks.

- Add toggle in privacy settings to enable/disable
- When enabled, route all evm broadcast rpc calls via Blink

Blazebrain committed:
feat: Handle same chainId scenario for pasted evm addresses

OmarHatem28 commented on cw_core/lib/erc20_token.dart:
```diff
@@ -74,9 +74,7 @@ class Erc20Token extends CryptoCurrency with HiveObjectMixin {
-  bool operator ==(other) =>
-      (other is Erc20Token && other.contractAddress == contractAddress) ||
-      (other is CryptoCurrency && other.title == title);
+  bool operator ==(Object other) => other is Erc20Token && other.contractAddress == contractAddress;
```
why is this change?

Blazebrain replied:
It affected the indexOf operator, which makes use of the "==" operator for comparison.

Based on the previous override, which adds `|| (other is CryptoCurrency && other.title == title);`, it means that any cryptocurrency that had the smae title compared the same, regardless of the chain/tag.

That's what caused one of the bugs Jack reported earlier, that USDC(Base) was returned, even though he selected USDC(Arb).

So indexOf stops at the first USDC(which was base), because it satisfies its check.

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
feat: Add Blink Setting to Advanced Settings Page

Blazebrain committed:
fix: Handle Blink supported chains

Blazebrain committed:
fix: Fixes to evm swap

Blazebrain committed:
fix: Add fallback for currency fetch

Blazebrain committed:
fix: Add evm to supported list

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
fix: Better handle balance errors

OmarHatem28 commented on lib/src/screens/wallet_connect/services/walletkit_service.dart:
```diff
@@ -562,9 +578,11 @@ abstract class WalletKitServiceBase with Store {
 
-        return isInCurrentTopics && isActive;
+        // return isInCurrentTopics && isActive;
+        return isInCurrentTopics;
```
why are those no longer needed?

Blazebrain replied:
Mainly because we no longer need them

For the isActive check, the pairing can be valid but inactive, and we want to be able to either disconnect properly or reconnect. The previous check filters out those types of pairing completely so we don't have the option to take action on them.

Same for the hasSession check, we have another way to retrieve sessions when needed, and a pairing should not depend on that.

Blazebrain committed:
fix: Update token check to use contract address comparison, resolves the usdt/usdt0 issue

Blazebrain committed:
refactor: Simplify checks by cearing all tokens on chain switch, removing unnecessary currency checks

Blazebrain commented on lib/src/screens/send/widgets/send_card.dart:
```diff
@@ -286,9 +301,27 @@ class SendCardState extends State<SendCard> with AutomaticKeepAliveClientMixin<S
+          sendViewModel.selectedCryptoCurrency = sendViewModel.wallet.currency;
+        } else {
+          sendViewModel.setSelectedCryptoCurrency(detectedCurrency.title);
+        }
```
Yeah, I think I initially added some other code into the if block earlier, and removed later on in the implementation.

But it can be called directly.

Blazebrain commented on lib/src/widgets/bottom_sheet/swap_confirmation_bottom_sheet.dart:
```diff
@@ -166,16 +167,20 @@ class SwapConfirmationContentState extends State<SwapConfirmationContent> {
-                            .iconPath!,
-                        width: 16,
-                        height: 16,
+                    if (isEVMCompatibleChain(widget.paymentFlowResult.walletType!)) ...[
```
It's used to show the network for the detected chain. 

So instead of just show USDC image alone, we show USDC image with the base network image shown at the bottom right, so the user knows its USDC on Base network.

Works like a badge. 

Other non-evm networks don't need this.

OmarHatem28 replied:
for me it seems like this is used to display the currency image, so now you are only showing it in case of an EVM wallet.

the if condition was not there, so this is not a new added badge, this hides an existing image in case of it's not evm, and I don't think we had the badge previously

Blazebrain commented on lib/view_model/exchange/exchange_view_model.dart:
```diff
@@ -185,6 +186,19 @@ abstract class ExchangeViewModelBase extends WalletChangeListenerViewModel with
+      receiveCurrency = wallet.currency;
+      depositCurrency = wallet.currency;
+      _injectUserEthTokensIntoCurrencyLists();
+    });
```
Yeah, updated to list to wallet currency

Blazebrain commented on lib/view_model/wallet_keys_view_model.dart:
```diff
@@ -14,25 +14,44 @@ import 'package:flutter/foundation.dart';
+              '${currentChain?.name ?? walletTypeToString(wallet.type)} ${S.current.wallet_keys}';
+        }
+      });
+    }
```
Yeah, the user can't switch chains here. The title updates here too, but I'll move it to another entry

Blazebrain committed:
fix: Simplify checks based from feedback on review

Blazebrain committed:
fix: Exclude arbitrum from tx priority using chainId

Blazebrain committed:
refactor: Handle chains not having transaction priority within refactor

Blazebrain committed:
fix: Unified evm wallet fixes
- Handle switching networks while on tx history page
- Update NFTs when network is switching on listing page
- Fix arbitrum tx priority error on swap screen
- Switch check to display deuro from wallet type to chainId
- Add fallback check for getting token info for evm
- Better error message for max fee per gas less than block base fee error

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
refactor: Update wallet checks from type to chainId for EVM compatibility checks

Blazebrain committed:
refactor: Update eth chain utils and remove verified contract criteria in scam detection checks

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
refactor: Enhance validation for deuro calls

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
fix : Remove unneeded validation check for savings edit

OmarHatem28 committed:
Merge branch 'main' into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
feat: Hide EVM chain switcher behind feature flag

Blazebrain committed:
refactor: Remove EVM wallet type references and update related logic

Blazebrain committed:
chore: Clean up wallet type references and remove repeated imports

OmarHatem28 commented on cw_evm/lib/evm_chain_wallet_service.dart:
```diff
@@ -1,56 +1,358 @@
+    final wallet = _createWalletInstance(
+      walletType: walletInfo.type,
+      walletInfo: walletInfo,
+      derivationInfo: await walletInfo.getDerivationInfo(),
```
at this point the sqlite table doesn't have the derivation info right? how would this work?
shouldn't this be something we get from the chainConfig?
or default it to `m/44'/60'/` since this is the derivation path for evm wallets?

Blazebrain replied:
Ohh, that's true. It will default to the unknown derivation type and no derivation type at that point.

I've added logic to set the default derivation path for EVM wallets if it's missing, so even if the table doesn't have it, we can still have the default to use for creating the wallet instance

OmarHatem28 commented on lib/view_model/settings/other_settings_view_model.dart:
```diff
@@ -67,7 +69,7 @@ abstract class OtherSettingsViewModelBase with Store {
   @computed
   bool get displayTransactionPriority =>
-      !(changeRepresentativeEnabled || [WalletType.solana, WalletType.tron].contains(_wallet.type));
+      !(changeRepresentativeEnabled || [WalletType.solana, WalletType.tron].contains(_wallet.type) || chainId == 42161);
```
```suggestion
      !(changeRepresentativeEnabled || [WalletType.solana, WalletType.tron, WalletType.arbitrum].contains(_wallet.type));
```

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
feat: Add fallback for derivation path and add arbitrum support to app scripts

Blazebrain committed:
fix: NFT popup error

Blazebrain committed:
fix: Transaction details fee regression

Blazebrain committed:
fix: Fixes and improvements
- Add nonce parameter to transaction methods in EVM clients
- Remove brackets for create swaps
- Display contacts for selected wallettype alone
- Remove tx priorty switching for arb

Blazebrain committed:
feat: Enhance Arbitrum support across swap providers

Blazebrain committed:
Merge branch 'main' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
Merge branch 'dev' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
Merge branch 'dev' into CW-1250-Refactor-L2-Chains-Code

Blazebrain committed:
fix: Add missing secret keys

Blazebrain committed:
Merge branch 'CW-1250-Refactor-L2-Chains-Code' of https://github.com/cake-tech/cake_wallet into CW-1250-Refactor-L2-Chains-Code

OmarHatem28 reviewed (approved):
Missing Arbitrum scheme in AndroidManifestBase and info.plist and adding this scheme to restore from qr flow

Blazebrain committed:
fix: display sending currency icon in send from external page

Blazebrain committed:
refactor: Handle arbitrum edgecase for wallet list display image

Blazebrain committed:
fix: Handle more cases of arbitrum image display for contact book

Blazebrain committed:
exploring possible issues with tx history

Blazebrain committed:
chore: switch to printV

Blazebrain committed:
fix: Add an extra check for keys

Blazebrain committed:
fix: Filter out eth spam transaction and modify parsing format

Blazebrain commented:
> Missing Arbitrum scheme in AndroidManifestBase and info.plist and adding this scheme to restore from qr flow

They are all added, or are you referring to something else?

Blazebrain committed:
Update local

Blazebrain committed:
fix: Handle contract decimals for sending tokens

Blazebrain committed:
fix: Display proper balance error

OmarHatem28 commented on cw_evm/lib/utils/evm_chain_formatter.dart:
```diff
@@ -0,0 +1,45 @@
+
+      final wholeBigInt = BigInt.parse(whole);
+      final fractionBigInt = BigInt.parse(fractionPadded);
+      final multiplier = BigInt.from(10).pow(decimals);
```
```suggestion
      String cleanAmount = amount.replaceAll(',', '.');

      bool isNegative = cleanAmount.startsWith('-');
      if (isNegative) {
        cleanAmount = cleanAmount.substring(1);
      }

      final parts = cleanAmount.split('.');
      String whole = parts[0].isEmpty ? '0' : parts[0];
      String fraction = parts.length > 1 ? parts[1] : '';

      // 3. Strict Truncation Logic
      // If fraction is longer than decimals, strictly cut it off.
      // If shorter, pad it with zeros.
      if (fraction.length > decimals) {
        fraction = fraction.substring(0, decimals);
      } else {
        fraction = fraction.padRight(decimals, '0');
      }

      final wholeBigInt = BigInt.parse(whole);
      final fractionBigInt = BigInt.parse(fraction);
      final multiplier = BigInt.from(10).pow(decimals);
```

OmarHatem28 committed:
Update cw_evm/lib/utils/evm_chain_formatter.dart [skip ci]
